File: components/credentials/src/test/java/org/cloudfoundry/credhub/repositories/CredentialVersionDataRepositoryTest.java
package org.cloudfoundry.credhub.repositories;

import java.util.Arrays;
import java.util.UUID;
import java.util.stream.Stream;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit4.SpringRunner;

import org.cloudfoundry.credhub.CredhubTestApp;
import org.cloudfoundry.credhub.entities.EncryptedValue;
import org.cloudfoundry.credhub.entities.EncryptionKeyCanary;
import org.cloudfoundry.credhub.entity.CertificateCredentialVersionData;
import org.cloudfoundry.credhub.entity.Credential;
import org.cloudfoundry.credhub.entity.ValueCredentialVersionData;
import org.cloudfoundry.credhub.utils.DatabaseProfileResolver;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import static java.nio.charset.StandardCharsets.UTF_8;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsEqual.equalTo;
import static org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase.Replace;

@RunWith(SpringRunner.class)
@ActiveProfiles(value = "unit-test", resolver = DatabaseProfileResolver.class)
@AutoConfigureTestDatabase(replace = Replace.NONE)
@SpringBootTest(classes = CredhubTestApp.class)
public class CredentialVersionDataRepositoryTest {

  @Autowired
  private CredentialVersionRepository subject;

  @Autowired
  private CredentialRepository credentialRepository;

  @Autowired
  private EncryptionKeyCanaryRepository canaryRepository;

  private String name;
  private UUID canaryUuid;

  @Before
  public void beforeEach() {
    name = "my-credential";
    final EncryptionKeyCanary canary = canaryRepository.save(new EncryptionKeyCanary());
    canaryUuid = canary.getUuid();
  }

  @After
  public void tearDown() {
    credentialRepository.deleteAll();
  }

  @Test
  public void canSaveCertificatesOfLength7000WhichMeans7016ForGCM() {
    final byte[] encryptedValue = new byte[7016];
    Arrays.fill(encryptedValue, (byte) 'A');
    final StringBuilder stringBuilder = new StringBuilder(7000);
    Stream.generate(() -> "a").limit(stringBuilder.capacity()).forEach(stringBuilder::append);

    final Credential credential = credentialRepository.save(new Credential(name));
    final String longString = stringBuilder.toString();

    final EncryptedValue entityEncryptedValue = new EncryptedValue();
    entityEncryptedValue.setEncryptionKeyUuid(canaryUuid);
    entityEncryptedValue.setEncryptedValue(encryptedValue);
    entityEncryptedValue.setNonce("nonce".getBytes(UTF_8));

    final CertificateCredentialVersionData entity = new CertificateCredentialVersionData();
    entity.setCredential(credential);
    entity.setCa(longString);
    entity.setCertificate(longString);
    entity.setEncryptedValueData(entityEncryptedValue);

    subject.save(entity);
    final CertificateCredentialVersionData credentialData = (CertificateCredentialVersionData) subject
      .findFirstByCredentialUuidOrderByVersionCreatedAtDesc(credential.getUuid());
    assertThat(credentialData.getCa().length(), equalTo(7000));
    assertThat(credentialData.getCertificate().length(), equalTo(7000));
    assertThat(credentialData.getEncryptedValueData().getEncryptedValue(), equalTo(encryptedValue));
    assertThat(credentialData.getEncryptedValueData().getEncryptedValue().length, equalTo(7016));
  }

  @Test
  public void canSaveStringsOfLength7000WhichMeans7016ForGCM() {
    final byte[] encryptedValue = new byte[7016];
    Arrays.fill(encryptedValue, (byte) 'A');

    final StringBuilder stringBuilder = new StringBuilder(7000);
    Stream.generate(() -> "a").limit(stringBuilder.capacity()).forEach(stringBuilder::append);
    final ValueCredentialVersionData entity = new ValueCredentialVersionData();

    final EncryptedValue entityEncryptedValue = new EncryptedValue();
    entityEncryptedValue.setEncryptedValue(encryptedValue);
    entityEncryptedValue.setEncryptionKeyUuid(canaryUuid);
    entityEncryptedValue.setNonce("nonce".getBytes(UTF_8));

    final Credential credential = credentialRepository.save(new Credential(name));
    entity.setCredential(credential);
    entity.setEncryptedValueData(entityEncryptedValue);

    subject.save(entity);
    EncryptedValue encryptedValueData = subject.findFirstByCredentialUuidOrderByVersionCreatedAtDesc(credential.getUuid()).getEncryptedValueData();
    assert encryptedValueData != null;
    assertThat(encryptedValueData.getEncryptedValue().length, equalTo(7016));
  }
}

File: components/credentials/src/test/kotlin/org/cloudfoundry/credhub/repositories/CredentialVersionDataRepositoryTest.kt
package org.cloudfoundry.credhub.repositories

import org.cloudfoundry.credhub.CredhubTestApp
import org.cloudfoundry.credhub.entities.EncryptedValue
import org.cloudfoundry.credhub.entities.EncryptionKeyCanary
import org.cloudfoundry.credhub.entity.CertificateCredentialVersionData
import org.cloudfoundry.credhub.entity.Credential
import org.cloudfoundry.credhub.entity.ValueCredentialVersionData
import org.cloudfoundry.credhub.utils.DatabaseProfileResolver
import org.hamcrest.MatcherAssert
import org.hamcrest.core.IsEqual
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import java.nio.charset.StandardCharsets
import java.util.Arrays
import java.util.UUID
import java.util.stream.Stream

@ActiveProfiles(value = ["unit-test"], resolver = DatabaseProfileResolver::class)
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@SpringBootTest(classes = [CredhubTestApp::class])
class CredentialVersionDataRepositoryTest {
    @Autowired
    private val subject: CredentialVersionRepository? = null

    @Autowired
    private val credentialRepository: CredentialRepository? = null

    @Autowired
    private val canaryRepository: EncryptionKeyCanaryRepository? = null
    private var name: String? = null
    private var canaryUuid: UUID? = null

    @BeforeEach
    fun beforeEach() {
        name = "my-credential"
        val canary = canaryRepository!!.save(EncryptionKeyCanary())
        canaryUuid = canary.uuid
    }

    @AfterEach
    fun tearDown() {
        credentialRepository!!.deleteAll()
    }

    @Test
    fun canSaveCertificatesOfLength7000WhichMeans7016ForGCM() {
        val encryptedValue = ByteArray(7016)
        Arrays.fill(encryptedValue, 'A'.code.toByte())
        val stringBuilder = StringBuilder(7000)
        Stream.generate { "a" }.limit(stringBuilder.capacity().toLong()).forEach { str: String? -> stringBuilder.append(str) }
        val credential = credentialRepository!!.save(Credential(name))
        val longString = stringBuilder.toString()
        val entityEncryptedValue = EncryptedValue()
        entityEncryptedValue.encryptionKeyUuid = canaryUuid
        entityEncryptedValue.encryptedValue = encryptedValue
        entityEncryptedValue.nonce = "nonce".toByteArray(StandardCharsets.UTF_8)
        val entity = CertificateCredentialVersionData("test-ca")
        entity.credential = credential
        entity.ca = longString
        entity.certificate = longString
        entity.setEncryptedValueData(entityEncryptedValue)
        subject!!.save(entity)
        val credentialData =
            subject
                .findFirstByCredentialUuidOrderByVersionCreatedAtDesc(credential.uuid) as CertificateCredentialVersionData?
        MatcherAssert.assertThat(credentialData!!.ca!!.length, IsEqual.equalTo(7000))
        MatcherAssert.assertThat(credentialData.certificate!!.length, IsEqual.equalTo(7000))
        MatcherAssert.assertThat(credentialData.getEncryptedValueData()!!.encryptedValue, IsEqual.equalTo(encryptedValue))
        MatcherAssert.assertThat(credentialData.getEncryptedValueData()!!.encryptedValue.size, IsEqual.equalTo(7016))
    }

    @Test
    fun canSaveStringsOfLength7000WhichMeans7016ForGCM() {
        val encryptedValue = ByteArray(7016)
        Arrays.fill(encryptedValue, 'A'.code.toByte())
        val stringBuilder = StringBuilder(7000)
        Stream.generate { "a" }.limit(stringBuilder.capacity().toLong()).forEach { str: String? -> stringBuilder.append(str) }
        val entity = ValueCredentialVersionData("test-credential")
        val entityEncryptedValue = EncryptedValue()
        entityEncryptedValue.encryptedValue = encryptedValue
        entityEncryptedValue.encryptionKeyUuid = canaryUuid
        entityEncryptedValue.nonce = "nonce".toByteArray(StandardCharsets.UTF_8)
        val credential = credentialRepository!!.save(Credential(name))
        entity.credential = credential
        entity.setEncryptedValueData(entityEncryptedValue)
        subject!!.save(entity)
        val encryptedValueData = subject.findFirstByCredentialUuidOrderByVersionCreatedAtDesc(credential.uuid)!!.getEncryptedValueData()!!
        MatcherAssert.assertThat(encryptedValueData.encryptedValue.size, IsEqual.equalTo(7016))
    }
}
